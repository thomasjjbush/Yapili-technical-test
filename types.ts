import { ReactNode } from 'react';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';
import { MockStoreEnhanced } from 'redux-mock-store';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface Character {
    comics?: CharecterCollection;
    description?: string;
    error?: boolean;
    events?: CharecterCollection;
    id?: number;
    modified?: string;
    name?: string;
    resourceURI?: string;
    series?: CharecterCollection;
    stories?: CharecterCollection;
    thumbnail?: {
        extension: string;
        path: string;
    };
}

export enum CharacterActions {
    CLEARED = 'character/CLEARED',
    ERROR = 'character/ERROR',
    LOADED = 'character/LOADED',
}

interface CharecterCollection {
    available?: number;
    collectionURI?: string;
    items: Resource[];
    returned?: number;
}

export enum Dimensions {
    LANDSCAPE = 'landscape',
    STANDARD = 'standard',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface FallbackImageProps {
    src: string;
    width?: string;
    height?: string;
}

export interface Home extends Partial<MarvelResponse> {
    error?: boolean;
    results?: Character[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LOADED = 'home/LOADED',
    LOADING = 'home/LOADING',
}

export interface ImageProps {
    dimension?: Dimensions;
    extension: string;
    height?: string;
    path: string;
    width?: string;
}

export interface MarvelResponse {
    count: number;
    limit: number;
    offset: number;
    results: any[];
    total: number;
}

export type MockStore = MockStoreEnhanced<Partial<Store>, Dispatch>;

export interface Options {
    [key: string]: string | number;
    apikey?: string;
    limit?: number;
    offset?: number;
}

export interface PaginatedButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface Resource {
    resourceURI: string;
    name: string;
    type?: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface Store {
    character: Character;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    breakpoints: {
        desktop: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        darkGrey: string;
        marvelRed: string;
        mediumGrey: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
