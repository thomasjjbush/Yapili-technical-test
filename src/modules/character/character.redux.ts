import { AnyAction } from 'redux';
import { createAction } from '../../utils/action/action';
import { db } from '../../utils/db/db';
import { Action, Character, CharacterActions } from '../../../types';

export const loadCharacter = (id: number): Action => async (dispatch, getState) => {
    try {
        dispatch(
            createAction(
                CharacterActions.LOADED,
                getState().home.results?.find((r) => r.id === id) ?? (await db(`characters/${id}`))?.results?.[0],
            ),
        );
    } catch (err) {
        dispatch(createAction(CharacterActions.ERROR));
    }
};

export const characterReducer = (state: Character = {}, { payload, type }: AnyAction): Character => {
    switch (type) {
        case CharacterActions.CLEARED:
            return {};
        case CharacterActions.ERROR:
            return { ...state, error: true };
        case CharacterActions.LOADED:
            return { ...state, ...payload };
        default:
            return state;
    }
};
