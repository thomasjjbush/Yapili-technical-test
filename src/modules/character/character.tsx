import React, { FC, ReactElement, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import styled from 'styled-components';
import { loadCharacter } from './character.redux';
import { Grid } from '../../components/grid/grid';
import { Image } from '../../components/image/image';
import {
    Character as CharacterS,
    CharacterActions,
    Dimensions,
    Resource as ResourceT,
    Store,
    StyledProps,
} from '../../../types';

const Container = styled.div`
    margin: auto;
    max-width: 800px;
    position: relative;

    img {
        margin-bottom: 30px;
    }
`;

const List = styled.ul`
    list-style-type: none;
    margin: 0;
    padding: 0;
`;

const ListItem = styled.li<StyledProps>`
    padding: 10px;
    &:not(:last-of-type) {
        border-bottom: solid 2px ${({ theme }) => theme.colors.marvelRed};
    }
`;

const Resource = styled.div`
    border: solid 2px #ffffff;
    border-radius: 5px;
`;

const renderResource = (name: string, items: ResourceT[]): ReactElement => (
    <Resource>
        <h1>{name}</h1>
        <List>
            {items.map(({ resourceURI, name }) => (
                <ListItem key={resourceURI}>{name}</ListItem>
            ))}
        </List>
    </Resource>
);

export const Character: FC = (): ReactElement => {
    const dispatch = useDispatch();
    const { comics, error, name, series, stories, thumbnail } = useSelector<Store, CharacterS>(
        (state) => state.character,
    );
    const { goBack } = useHistory();
    const { id } = useParams<{ id: string }>();

    useEffect(() => {
        dispatch(loadCharacter(Number(id)));
        return () => dispatch({ type: CharacterActions.CLEARED });
    }, [id]);

    if (error) return <p>error!</p>;
    if (!name) return <p>loading!</p>;

    return (
        <Container>
            <h1>{name}</h1>
            {thumbnail && (
                <Image
                    dimension={Dimensions.LANDSCAPE}
                    extension={thumbnail.extension}
                    path={thumbnail.path}
                    width="100%"
                />
            )}
            <Grid contentHeight desktop={3} gap={30} mobile={1} tablet={2}>
                {renderResource('Comics', comics.items)}
                {renderResource('Series', series.items)}
                {renderResource('Stories', stories.items)}
            </Grid>
            <button aria-label="Go back" onClick={goBack} type="button">
                üè†
            </button>
        </Container>
    );
};
