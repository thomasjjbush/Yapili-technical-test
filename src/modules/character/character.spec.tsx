import React, { ReactElement } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { Character } from './character';
import { Provider } from 'react-redux';
import { BrowserRouter, Route } from 'react-router-dom';
import { mockStore, mockTheme } from '../../../test/jest.utils';
import { loadCharacter } from './character.redux';
import { Character as CharacterT } from '../../../types';
import { ThemeProvider } from 'styled-components';

delete global.window.location;
(global.window.location as any) = new URL('https://www.example.com/test/500');

const d = jest.fn();
const g = jest.fn();
jest.mock('react-redux', () => ({ ...jest.requireActual('react-redux'), useDispatch: jest.fn(() => d) }));
jest.mock('react-router-dom', () => ({ ...jest.requireActual('react-router-dom'), useHistory: () => ({ goBack: g }) }));
jest.mock('./character.redux', () => ({ loadCharacter: jest.fn().mockReturnValue({ type: 'MOCK_ACTION' }) }));
jest.mock('../../components/grid/grid', () => ({ Grid: ({ children }: { children: ReactElement[] }) => children }));
jest.mock('../../components/image/image', () => ({ Image: () => 'Image' }));

const ConnectedComponent = (character?: CharacterT): ReactWrapper =>
    mount(
        <Provider store={mockStore({ character })}>
            <ThemeProvider theme={mockTheme}>
                <BrowserRouter>
                    <Route path="/test/:id">
                        <Character />
                    </Route>
                </BrowserRouter>
            </ThemeProvider>
        </Provider>,
    );

describe('Character', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should return loading if no data is available', () => {
        expect(ConnectedComponent().find(Character).find('p').text()).toBe('loading!');
    });

    it('should return error if data includes an error', () => {
        expect(ConnectedComponent({ error: true }).find(Character).find('p').text()).toBe('error!');
    });

    it('should invoke load character', () => {
        ConnectedComponent();
        expect(loadCharacter).toHaveBeenCalledTimes(1);
        expect(loadCharacter).toHaveBeenCalledWith(500);
        expect(d).toHaveBeenNthCalledWith(1, { type: 'MOCK_ACTION' });
    });

    describe('with data', () => {
        const store = {
            comics: { items: [{ resourceURI: '1', name: 'comic' }] },
            name: 'name',
            series: { items: [{ resourceURI: '1', name: 'comic' }] },
            stories: { items: [{ resourceURI: '1', name: 'comic' }] },
            thumbnail: { path: 'path/', extension: 'ext' },
        };

        it('should match snapshot if data exists', () => {
            const wrapper = ConnectedComponent(store);
            expect(wrapper.find(Character)).toMatchSnapshot();
        });

        it('should be able to navigate home on click', () => {
            const button = ConnectedComponent(store).find('button');
            button.simulate('click');
            expect(g).toHaveBeenCalledTimes(1);
        });
    });
});
