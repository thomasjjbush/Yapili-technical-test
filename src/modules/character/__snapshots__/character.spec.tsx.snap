// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character with data should match snapshot if data exists 1`] = `
<Character>
  <styled.div
    styles="
    margin: auto;
    max-width: 800px;
    position: relative;

    img {
        margin-bottom: 30px;
    }
"
  >
    <div
      className="sc-bdfBwQ jlLvgZ"
    >
      <h1>
        name
      </h1>
      <Image
        dimension="landscape"
        extension="ext"
        path="path/"
        width="100%"
      >
        Image
      </Image>
      <Grid
        contentHeight={true}
        desktop={3}
        gap={30}
        mobile={1}
        tablet={2}
      >
        <styled.div
          styles="
    border: solid 2px #ffffff;
    border-radius: 5px;
"
        >
          <div
            className="sc-hKgILt iiApof"
          >
            <h1>
              Comics
            </h1>
            <styled.ul
              styles="
    list-style-type: none;
    margin: 0;
    padding: 0;
"
            >
              <ul
                className="sc-gsTCUz lbbSTx"
              >
                <styled.li
                  key="1"
                  styles="
    padding: 10px;
    &:not(:last-of-type) {
        border-bottom: solid 2px theme.colors.marvelRed;
    }
"
                >
                  <li
                    className="sc-dlfnbm hiNrMd"
                  >
                    comic
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </div>
        </styled.div>
        <styled.div
          styles="
    border: solid 2px #ffffff;
    border-radius: 5px;
"
        >
          <div
            className="sc-hKgILt iiApof"
          >
            <h1>
              Series
            </h1>
            <styled.ul
              styles="
    list-style-type: none;
    margin: 0;
    padding: 0;
"
            >
              <ul
                className="sc-gsTCUz lbbSTx"
              >
                <styled.li
                  key="1"
                  styles="
    padding: 10px;
    &:not(:last-of-type) {
        border-bottom: solid 2px theme.colors.marvelRed;
    }
"
                >
                  <li
                    className="sc-dlfnbm hiNrMd"
                  >
                    comic
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </div>
        </styled.div>
        <styled.div
          styles="
    border: solid 2px #ffffff;
    border-radius: 5px;
"
        >
          <div
            className="sc-hKgILt iiApof"
          >
            <h1>
              Stories
            </h1>
            <styled.ul
              styles="
    list-style-type: none;
    margin: 0;
    padding: 0;
"
            >
              <ul
                className="sc-gsTCUz lbbSTx"
              >
                <styled.li
                  key="1"
                  styles="
    padding: 10px;
    &:not(:last-of-type) {
        border-bottom: solid 2px theme.colors.marvelRed;
    }
"
                >
                  <li
                    className="sc-dlfnbm hiNrMd"
                  >
                    comic
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </div>
        </styled.div>
      </Grid>
      <button
        aria-label="Go back"
        onClick={[MockFunction]}
        type="button"
      >
        üè†
      </button>
    </div>
  </styled.div>
</Character>
`;
